### PR TITLE
Djangofixes

### PR DESCRIPTION
**Pull request by sum12**
fixed to be able to use https://github.com/sum12/rqlite-django backend

### CODE REVIEW

**Comment by otoolep** [comment]
@zmedico -- any comment?

File: src/pyrqlite/cursors.py [hunk 1]
```diff
@@ -148,6 +150,10 @@ def _substitute_params(self, operation, parameters):
     def _get_sql_command(self, sql_str):
         return sql_str.split(None, 1)[0].upper()
 
+    def convert_query(self, query):
+        FORMAT_QMARK_REGEX = re.compile(r'(?<!%)%s')
```
Comment (zmedico) [inline src/pyrqlite/cursors.py]: This isn't safe, the parameter style really needs to be adjustable in the ORM, for example sqlalchemy's SQLiteDialect declares `default_paramstyle = 'qmark'` here:

https://github.com/zzzeek/sqlalchemy/blob/rel_1_2_10/lib/sqlalchemy/dialects/sqlite/base.py#L1043

**Review by zmedico — CHANGES_REQUESTED** [review]

File: src/pyrqlite/cursors.py [hunk 2]
```diff
@@ -156,11 +162,14 @@ def execute(self, operation, parameters=None):
         operation = self._substitute_params(operation, parameters)
 
         command = self._get_sql_command(operation)
+        transact = ''
```
Comment (zmedico) [inline src/pyrqlite/cursors.py]: When automatic transactions are disabled, it probably makes sense to support the `isolation_level` parameter that python's sqlite3 module supports:

https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.isolation_level

The default behavior, `isolation_level=''`, means to implicitly issue a plain BEGIN statement before a DML statement. The parameter can be set to DEFERRED, IMMEDIATE, or EXCLUSIVE in order to make it implicitly issue the corresponding type of BEGIN statement before a DML statement. For `isolation_level=None`, autocommit mode is triggered, as described here:

https://www.sqlite.org/c3ref/get_autocommit.html

File: src/pyrqlite/connections.py [hunk 3]
```diff
@@ -43,7 +43,7 @@ class Connection(object):
     def __init__(self, host='localhost', port=4001,
                  user=None, password=None, connect_timeout=None,
                  detect_types=0, max_redirects=UNLIMITED_REDIRECTS,
-                 always_transactions=True):
+                 always_transactions=True, default_paramstyle='qmark'):
```
Comment (zmedico) [inline src/pyrqlite/connections.py]: In my remark about default_paramstyle in sqlalchemy's SQLiteDialect, my intention was to suggest that the ORM conform to the paramstyle of the database. The existing convention is for paramstyle to be constant for a particular database type, see:

https://www.python.org/dev/peps/pep-0249/#paramstyle

**Comment by sum12** [comment]
so iiuc I can drop this change of param style, because django uses `qmark` style  for sqlite ? and that is already handled

and for the `isolation_level` iiuc, django would prefer `isolation_level=''` since it handles transactions by itself, but `pyrqlite` will have to do the right thing for other values of `isolation_level`. Did I get that right ?

**Comment by zmedico** [comment]
> so iiuc I can drop this change of param style, because django uses qmark style for sqlite ? and that is already handled

Yes, if django respects the `paramstyle` value defined in src/pyrqlite/dbapi2.py then it should behave correctly without any paramstyle changes in pyrqlite.

> and for the isolation_level iiuc, django would prefer isolation_level='' since it handles transactions by itself, but pyrqlite will have to do the right thing for other values of isolation_level. Did I get that right ?

If django works correctly with the default `isolation_level=''` setting of the sqlite3 module, then it should work fine with pyrqlite too if we implement the `isolation_level` parameter.

**Comment by sum12** [comment]
I have reverted most of my changes. Thanks to @zmedico for the patiently explaining stuff. The changes related to transaction and `isolation_level` are not required after turning off a "feature flag" `autocommits_when_autocommit_is_off` in the sqlite3 backend of django

edit: @zmedico if the changes look good to you, then let me know I will squash them according

File: src/pyrqlite/cursors.py [hunk 4]
```diff
@@ -277,8 +277,12 @@ def fetchone(self):
 
     def fetchmany(self, size=None):
         if size is None:
-            size = self.arraysize
-        raise NotImplementedError(self)
+            size = max(self.arraysize, len(self._rows))
+        size = min(len(self._rows), size)
```
Comment (zmedico) [inline src/pyrqlite/cursors.py:281]: The size calculation looks like in could be incorrect for some corner cases. Also, the [fetchmany documentation](https://www.python.org/dev/peps/pep-0249/#fetchmany) says to use arraysize when size has not been specified, so I'd like to conform to that. How about something like this:
```python
    def fetchmany(self, size=None):
        remaining = self.arraysize if size is None else size
        remaining = min(remaining, self.rowcount - self.rownumber)
        for i in range(remaining):
            rows.append(self.fetchone())
        return rows
```

Comment (zmedico) [reply src/pyrqlite/cursors.py:281]: I've merged the Row slice change in aa562e4942531e96a13b927b69be4bdf9ed6cc43, and the cursor factory change in 9da1eb12c66e73b425b70523984517ec744d3a5a.

**Comment by zmedico** [comment]
Changes merged in aa562e4942531e96a13b927b69be4bdf9ed6cc43, 9da1eb12c66e73b425b70523984517ec744d3a5a, and f8cd41e0522af36a234088b3c876ccfc5078f662. Thanks!

**Pull request closed without merging**